### PR TITLE
ASIM User Management schema parser with its sample and test data for SentinelOne

### PR DESCRIPTION
**Pull request by jayeshprajapaticrest**
Required items, please complete
   
   Change(s):
   - Added ASIM User Management schema parser for SentinelOne based on API Data

   Reason for Change(s):
   - Initial version ASIM User Management Parser

   Version Updated:
   - Initial version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

**Comment by v-sudkharat** [comment]
Hello @jayeshprajapaticrest, thank you for raising Pull Request with us! We will review it internally and get back to you by shortly.

**Comment by vakohl** [comment]
@jayeshprajapaticrest Will try to review this within this week

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      | extend 
+          ActorUsername = iff(activityType_d == 102, "SentinelOne", coalesce(byUser, username, email)),        
+          GroupName = coalesce(group, groupName, name),
+          TargetUsername = iff(isnotempty(byUser), username, ""),
+          PreviousPropertyValue = coalesce(oldDescription, oldRole),
+          NewPropertyValue = coalesce(description, role)
+      | extend GroupName = iff(GroupName == "null", "", GroupName)
+      | project-rename
+          EventStartTime = createdAt_t,
+          DvcIpAddr = ipAddress,
```
Comment (vakohl) [inline Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:68]: https://learn.microsoft.com/en-us/azure/sentinel/normalization-schema-user-management#source-fields
please check if we can use Src fields

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 2]
```diff
@@ -0,0 +1,134 @@
+Parser:
+  Title: User Management ASIM parser for SentinelOne
+  Version: '0.1.1'
```
Comment (vakohl) [inline Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:3]: 0.1.0 => Initial version

**Review by vakohl — CHANGES_REQUESTED** [review]

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, 
I would appreciate it if you could check for some changes @vakohl requested. Thanks

**Comment by jayeshprajapaticrest** [comment]
> Hi @jayeshprajapaticrest, I would appreciate it if you could check for some changes @vakohl requested. Thanks

@v-atulyadav 
Yes we are working on it. You can get update by Monday EOD.

File: Parsers/ASimUserManagement/Tests/SentinelOne_ASimUserManagement_SchemaTest.csv [hunk 3]
```diff
@@ -0,0 +1,41 @@
+﻿Result
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Tests/SentinelOne_ASimUserManagement_SchemaTest.csv:1]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  let EventTypeLookup = datatable (
+      activityType_d: real,
+      EventType: string,
+      EventOriginalType: string,
+      EventSubType: string
+  )[
+      23, "UserCreated", "User Added", "",
+      24, "UserModified", "User Modified", "MultipleProperties",
+      25, "UserDeleted", "User Deleted", "",
+      37, "UserModified", "User modified", "UserModified",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+          ActorUsername = iff(activityType_d == 102, "SentinelOne", coalesce(byUser, username, email)),        
+          GroupName = coalesce(group, groupName, name),
+          TargetUsername = iff(isnotempty(byUser), username, ""),
+          PreviousPropertyValue = coalesce(oldDescription, oldRole),
+          NewPropertyValue = coalesce(description, role)
+      | extend GroupName = iff(GroupName == "null", "", GroupName)
+      | project-rename
+          EventStartTime = createdAt_t,
+          DvcIpAddr = ipAddress,
+          EventUid = _ResourceId,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+          TargetUsernameType = _ASIM_GetUsernameType(TargetUsername),
+          AdditionalFields = bag_pack(
+                        "userScope", userScope,
+                        "scopeLevelName", scopeLevelName,
+                        "scopeName", scopeName,
+                        "modifiedFields", modifiedFields,
+                        "roleName", roleName,
+                        "deactivationPeriodInDays", deactivationPeriodInDays,
+                        "descriptionChanged", descriptionChanged,
+                        "groupType", groupType
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+          UpdatedPropertyName = EventSubType,
+          ActorUsernameType = _ASIM_GetUsernameType(ActorUsername),
+          GroupIdType = iff(isnotempty(GroupId), "UID", ""),
+          GroupNameType = iff(isnotempty(GroupName), "Simple", ""),
+          TargetUsernameType = _ASIM_GetUsernameType(TargetUsername),
+          AdditionalFields = bag_pack(
+                        "userScope", userScope,
+                        "scopeLevelName", scopeLevelName,
+                        "scopeName", scopeName,
+                        "modifiedFields", modifiedFields,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:105]: This field may have multiple values that's why we set UpdatedPropertyName(Alias of EventSubType) to MultipleProperties.

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+      | where event_name_s == "Activities."
+          and activityType_d in (23, 24, 25, 37, 102, 110, 111, 140, 141, 142, 3522, 3523, 3524, 3710, 3711, 3715, 5006, 5008, 5011)
+      | parse-kv DataFields_s as (byUser: string, username: string, email: string, ipAddress: string, group: string, groupName: string, name: string, oldDescription: string, oldRole: string, description: string, role: string, userScope: string, scopeLevelName: string, scopeName: string, roleName: string, modifiedFields: string, deactivationPeriodInDays: string, descriptionChanged: string, groupType: string) with (pair_delimiter=",", kv_delimiter=":", quote='"')
+      | parse description_s with * "with id=" id: string "," restOfMessage
+      | lookup EventTypeLookup on activityType_d
+      | extend 
+          ActorUsername = iff(activityType_d == 102, "SentinelOne", coalesce(byUser, username, email)),        
+          GroupName = coalesce(group, groupName, name),
+          TargetUsername = iff(isnotempty(byUser), username, ""),
+          PreviousPropertyValue = coalesce(oldDescription, oldRole),
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:65]: Yes

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+          NewPropertyValue = coalesce(description, role)
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:66]: Yes

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+          EventResultDetails = "Other"
+      | extend
+          Dvc = coalesce(DvcIpAddr, "SentinelOne"),
+          EventEndTime = EventStartTime,
+          User = ActorUsername,
+          UpdatedPropertyName = EventSubType,
+          ActorUsernameType = _ASIM_GetUsernameType(ActorUsername),
+          GroupIdType = iff(isnotempty(GroupId), "UID", ""),
+          GroupNameType = iff(isnotempty(GroupName), "Simple", ""),
+          TargetUsernameType = _ASIM_GetUsernameType(TargetUsername),
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:99]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+          GroupName = coalesce(group, groupName, name),
+          TargetUsername = iff(isnotempty(byUser), username, ""),
+          PreviousPropertyValue = coalesce(oldDescription, oldRole),
+          NewPropertyValue = coalesce(description, role)
+      | extend GroupName = iff(GroupName == "null", "", GroupName)
+      | project-rename
+          EventStartTime = createdAt_t,
+          DvcIpAddr = ipAddress,
+          EventUid = _ResourceId,
+          ActorUserId = id,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:72]: ActorUserId

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+          EventSchemaVersion = "0.1.1",
+          EventProduct = "SentinelOne",
+          EventVendor = "SentinelOne",
+          EventResultDetails = "Other"
+      | extend
+          Dvc = coalesce(DvcIpAddr, "SentinelOne"),
+          EventEndTime = EventStartTime,
+          User = ActorUsername,
+          UpdatedPropertyName = EventSubType,
+          ActorUsernameType = _ASIM_GetUsernameType(ActorUsername),
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:94]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml]: Yes. No other Dvc fields are available in the SentinelOne data.

Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml (hunk 2 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:72]: Done

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 13]
```diff
@@ -0,0 +1,134 @@
+Parser:
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:1]: Done. Added

File: Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  Title: User Management ASIM parser for SentinelOne
+  Version: '0.1.0'
+  LastUpdated: Jul 25, 2023
```
Comment (vakohl) [inline Parsers/ASimUserManagement/Parsers/ASimUserManagementSentinelOne.yaml:4]: please keep same update date as vim parser

File: Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+          EventMessage = primaryDescription_s,
+          EventOriginalUid = activityUuid_g
+      | extend
+          EventCount = int(1),
+          EventResult = "Success",
+          DvcAction = "Allowed",
+          EventSeverity = "Informational",
+          EventSchema = "UserManagement",
+          EventSchemaVersion = "0.1.1",
+          EventProduct = "SentinelOne",
```
Comment (vakohl) [inline Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml:116]: Please add SentinelOne as supported value for EventVendor and EventProduct

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml (hunk 15 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml:116]: Done

File: Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+  };
+  parser(
+    starttime = starttime,
+    endtime = endtime,
+    targetusername_has = targetusername_has,
+    actorusername_has = actorusername_has,
+    targetdomain_has = targetdomain_has,
+    anydomain_has = anydomain_has,
+    disabled=disabled
+  )
```
Comment (vakohl) [inline Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml:185]: Team, no threat fields found for user management?

Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml:185]: No threat field available for user related data in sentinel.

File: Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+    Default: '*'
+  - Name: targetdomain_has
+    Type: string
+    Default: '*'
+  - Name: anydomain_has
+    Type: string    
+    Default: '*'
+  - Name: disabled
+    Type: bool
+    Default: false
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimUserManagement/Parsers/vimUserManagementSentinelOne.yaml:42]: Done

**Review by vakohl — APPROVED** [review]

File: ASIM/dev/ASimTester/ASimTester.csv [hunk 18]
```diff
@@ -554,7 +554,7 @@ EventProductVersion,string,Optional,FileEvent,,,
 EventProductVersion,string,Optional,NetworkSession,,,
 EventProductVersion,string,Optional,ProcessEvent,,,
 EventProductVersion,string,Optional,RegistryEvent,,,
-EventProductVersion,string,Optional,UserManagement,,,
+EventProductVersion,string,Optional,UserManagement,Enumerated,SentinelOne,
```
Comment (vakohl) [inline ASIM/dev/ASimTester/ASimTester.csv:557]: @jayeshprajapaticrest 
Please correct this

Comment (vakohl) [reply ASIM/dev/ASimTester/ASimTester.csv:557]: pls double check this in other parsers

**Comment by v-atulyadav** [comment]
hi @jayeshprajapaticrest, 
I have resolved branch conflicts, please see @vakohl's comment. Thanks

File: ASIM/dev/ASimTester/ASimTester.csv (hunk 18 above)
Comment (jayeshprajapaticrest) [reply ASIM/dev/ASimTester/ASimTester.csv]: Done

**Comment by jayeshprajapaticrest** [comment]
> hi @jayeshprajapaticrest, I have resolved branch conflicts, please see @vakohl's comment. Thanks

Addressed.

**Review by vakohl — APPROVED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**